process-engine/documentation
# ✨ embed spectacle-docs rendered api-docs

## What did you change?

I was researching a way to render an OpenAPI-formatted API-Documentation as static document and embed it in our gitbook. This is my POC-Version of a possible solution.

I decided to use [spectacle-docs](https://github.com/sourcey/spectacle) as entry-point. It allows to render a static html-document out of an API-Documentation. It is already highly customizable out-of-the-box.

I faced a couple of issues:
1. Gitbook really likes white space left and right of the main-content-area
2. Gitbook does *not* like loading local js-files
3. Although highly customizable, you can not add build-steps to spectacle out-of the box

My solutions in this branch:
1. customize the default gitbook-layout and add a couple of css-classes so that pages that start
  with an iframe will have that iframe fill the entire page. gitbook allows for easy customization
  of these things.

2. I added a custom build-step to spectacle that allows me to add the necessary js and css
  file-contents to the handlebar-template-data and therefore render them inline to the html-file
  instead of referencing the files

3. I copied the spectacle-program (the core-program only consists of two files),
  aggregated it to a `spectacle.js` and added that to the repo. by running this instead of the
  regular spectacle i could add custom tasks

Because the whole document-rendering-software and all customization to it and the gitbook are in
this repo now we have 100% control over the output and can completely customize it to our hearts
content.

A more detailed description of what i did and why is in the [apidoc.md](https://github.com/process-engine/documentation/blob/32c7a551866c3b3ebab2d7ef207f4ab9f05a1453/apidoc.md)

That description is in german and the code is not optimal yet because this is still in POC-phase, but it allows us to actually do the REST-API-documentation. This PR is so huge because the whole spectacle-program is now in the repo

![bildschirmfoto 2017-11-23 um 12 01 34](https://user-images.githubusercontent.com/20770029/33169774-1f8c827c-d046-11e7-84fb-527c3293835e.png)

## How can others test the changes?

- check out the branch and run `npm install`
- run `npm run render_api_doc`
- copy the now existing `public/index.html` to the not yet existing `reference/apidoc.html`
- run `npm start`
- visit `http://localhost:4000/reference/` and see the api-doc that was rendered from the `swagger.json`

## PR-Checklist

Please check the boxes in this list after submitting your PR:
